dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/DefaultJpaQueryParser.java

```diff
@@ -137,43 +149,43 @@ private Restriction getRestriction(
       case "endsWith", "ilike$" ->
           Restrictions.ilike(path, parseValue(valueType, arg), MatchMode.END);
       case "!ilike$" -> Restrictions.notIlike(path, parseValue(valueType, arg), MatchMode.END);
-      case "in" -> {
-        Collection values = null;
-
-        if (property.isCollection()) {
-          values = parseValue(Collection.class, property.getItemKlass(), arg);
-        } else {
-          values = parseValue(Collection.class, valueType, arg);
-        }
-
-        if (values == null || values.isEmpty()) {
-          throw new QueryParserException("Invalid argument `" + arg + "` for in operator.");
-        }
-
-        yield Restrictions.in(path, values);
-      }
-      case "!in" -> {
-        Collection values = null;
-
-        if (property.isCollection()) {
-          values = parseValue(Collection.class, property.getItemKlass(), arg);
-        } else {
-          values = parseValue(Collection.class, valueType, arg);
-        }
-
-        if (values == null || values.isEmpty()) {
-          throw new QueryParserException("Invalid argument `" + arg + "` for in operator.");
-        }
-
-        yield Restrictions.notIn(path, values);
-      }
+      case "in" -> Restrictions.in(path, parseValues(property, valueType, arg));
+      case "!in" -> Restrictions.notIn(path, parseValues(property, valueType, arg));
       case "null" -> Restrictions.isNull(path);
       case "!null" -> Restrictions.isNotNull(path);
       case "empty" -> Restrictions.isEmpty(path);
       default -> throw new QueryParserException("`" + operator + "` is not a valid operator.");
     };
   }
 
+  @Nonnull
+  @SuppressWarnings("rawtypes")
+  private Collection parseValues(Property property, Class<?> valueType, Object arg) {
+    Collection<?> values =
+        property.isCollection()
+            ? parseValue(Collection.class, property.getItemKlass(), arg)
+            : parseValue(Collection.class, valueType, arg);
+
+    if (values == null || values.isEmpty()) {
+      throw new QueryParserException("Invalid argument `" + arg + "` for in operator.");
+    }
+    return values;
+  }
+
+  private Collection<Disjunction> getDisjunctionsFromCustomMentions(
```

Comment:
FYI: logic extracted from outside of the engine to be inside as it operates on the 3 parts on a filter